File: data-pipeline/ep-core/src/main/java/org/ekstep/ep/samza/util/RedisConnect.java
package org.ekstep.ep.samza.util;

import org.apache.samza.config.Config;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import redis.clients.jedis.exceptions.JedisDataException;
import redis.clients.jedis.exceptions.JedisException;
import com.google.gson.Gson;

import java.time.Duration;


public class RedisConnect {
    private JedisPool jedisPool;
    private Config config;
    private Gson gson = new Gson();

    public RedisConnect(Config config) {
        this.config = config;
        String redis_host = config.get("redis.host", "localhost");
        Integer redis_port = config.getInt("redis.port", 6379);
        this.jedisPool = new JedisPool(buildPoolConfig(), redis_host, redis_port);
    }

    private JedisPoolConfig buildPoolConfig() {
        final JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setMaxTotal(config.getInt("redis.connection.max", 2));
        poolConfig.setMaxIdle(config.getInt("redis.connection.idle.max", 2));
        poolConfig.setMinIdle(config.getInt("redis.connection.idle.min", 1));
        poolConfig.setTestOnBorrow(true);
        poolConfig.setTestOnReturn(true);
        poolConfig.setTestWhileIdle(true);
        poolConfig.setMinEvictableIdleTimeMillis(Duration.ofSeconds(config.getInt("redis.connection.minEvictableIdleTimeSeconds", 120)).toMillis());
        poolConfig.setTimeBetweenEvictionRunsMillis(Duration.ofSeconds(config.getInt("redis.connection.timeBetweenEvictionRunsSeconds", 300)).toMillis());
        poolConfig.setNumTestsPerEvictionRun(3);
        poolConfig.setBlockWhenExhausted(true);
        return poolConfig;
    }

    public Jedis getConnection() {
        return jedisPool.getResource();
    }

    public Jedis getConnection(int database) {
        Jedis conn = jedisPool.getResource();
        conn.select(database);
        return conn;
    }

    public void resetConnection() {
        this.jedisPool.close();
        String redis_host = config.get("redis.host", "localhost");
        Integer redis_port = config.getInt("redis.port", 6379);
        this.jedisPool = new JedisPool(buildPoolConfig(), redis_host, redis_port);
    }

    public Jedis resetConnection(int database) {
        resetConnection();
        return getConnection(database);
    }

    public void addToCache(String key, String value, Jedis redisConnection, int storeId) {
        try {
            if (key != null && !key.isEmpty() && null != value && !value.isEmpty()) {
                redisConnection.set(key, value);
            }
        } catch(JedisException ex) {
            redisConnection = resetConnection(storeId);
            if (null != value)
                redisConnection.set(key, value);
        }
    }

    public String readFromCache(String key, Jedis redisConnection, int storeId) {
        try {
            return redisConnection.get(key);
        }
        catch (JedisException ex) {
            redisConnection = resetConnection(storeId);
            return redisConnection.get(key);
        }
    }
}

File: data-pipeline-flink/dp-core/src/main/scala/org/ekstep/dp/cache/RedisConnect.scala
package org.ekstep.dp.cache

import redis.clients.jedis.Jedis
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPoolConfig
import java.time.Duration

import com.typesafe.config.Config
import org.ekstep.dp.core.BaseJobConfig
import org.slf4j.LoggerFactory


class RedisConnect(jobConfig: BaseJobConfig) extends java.io.Serializable {

  private val serialVersionUID = - 396824011996012513L

  val config: Config = jobConfig.config
  val redis_host: String = Option(config.getString("redis.host")).getOrElse("localhost")
  val redis_port: Int = Option(config.getInt("redis.port")).getOrElse(6379)
  private var jedisPool = new JedisPool(buildPoolConfig, redis_host, redis_port)
  private val logger = LoggerFactory.getLogger(classOf[RedisConnect])

  private def buildPoolConfig = {
    val poolConfig = new JedisPoolConfig
    poolConfig.setMaxTotal(Option(config.getInt("redis.connection.max")).getOrElse(2))
    poolConfig.setMaxIdle(Option(config.getInt("redis.connection.idle.max")).getOrElse(2))
    poolConfig.setMinIdle(Option(config.getInt("redis.connection.idle.min")).getOrElse(1))
    poolConfig.setTestOnBorrow(true)
    poolConfig.setTestOnReturn(true)
    poolConfig.setTestWhileIdle(true)
    poolConfig.setMinEvictableIdleTimeMillis(Duration.ofSeconds(Option(config.getLong("redis.connection.minEvictableIdleTimeSeconds")).getOrElse(120L)).toMillis)
    poolConfig.setTimeBetweenEvictionRunsMillis(Duration.ofSeconds(Option(config.getLong("redis.connection.timeBetweenEvictionRunsSeconds")).getOrElse(300)).toMillis)
    poolConfig.setNumTestsPerEvictionRun(3)
    poolConfig.setBlockWhenExhausted(true)
    poolConfig
  }

  def getConnection: Jedis = {
    logger.info("Obtaining new Redis connection...")
    jedisPool.getResource
  }

  def getConnection(database: Int): Jedis = {
    val conn = jedisPool.getResource
    conn.select(database)
    conn
  }

  def resetConnection(): Unit = {
    this.jedisPool.close()
    val redis_host = Option(config.getString("redis.host")).getOrElse("localhost")
    val redis_port = Option(config.getInt("redis.port")).getOrElse(6379)
    this.jedisPool = new JedisPool(buildPoolConfig, redis_host, redis_port)
  }

  def resetConnection(database: Int): Jedis = {
    resetConnection()
    getConnection(database)
  }
}
